src/Generics/SOP/BasicFunctors.hs

```diff
@@ -108,14 +197,81 @@ unI (I x) = x
 -- and with a shorter name.
 --
 newtype (:.:) (f :: l -> *) (g :: k -> l) (p :: k) = Comp (f (g p))
-  deriving (Show, GHC.Generic)
+  deriving (GHC.Generic)
 
 infixr 7 :.:
 
 instance (Functor f, Functor g) => Functor (f :.: g) where
   fmap f (Comp x) = Comp (fmap (fmap f) x)
 
+-- Instances of lifted Prelude classes
+
+#ifdef LIFTED_CLASSES
+instance (Eq1 f, Eq1 g) => Eq1 (f :.: g) where
+    liftEq eq (Comp x) (Comp y) = liftEq (liftEq eq) x y
+
+instance (Ord1 f, Ord1 g) => Ord1 (f :.: g) where
+    liftCompare comp (Comp x) (Comp y) =
+        liftCompare (liftCompare comp) x y
+
+instance (Read1 f, Read1 g) => Read1 (f :.: g) where
+    liftReadsPrec rp rl = readsData $
+        readsUnaryWith (liftReadsPrec rp' rl') "Comp" Comp
+      where
+        rp' = liftReadsPrec rp rl
+        rl' = liftReadList rp rl
+
+instance (Show1 f, Show1 g) => Show1 (f :.: g) where
+    liftShowsPrec sp sl d (Comp x) =
+        showsUnaryWith (liftShowsPrec sp' sl') "Comp" d x
+      where
+        sp' = liftShowsPrec sp sl
+        sl' = liftShowList sp sl
+
+instance (Eq1 f, Eq1 g, Eq a) => Eq ((f :.: g) a) where (==) = eq1
+instance (Ord1 f, Ord1 g, Ord a) => Ord ((f :.: g) a) where compare = compare1
+instance (Read1 f, Read1 g, Read a) => Read ((f :.: g) a) where readsPrec = readsPrec1
+instance (Show1 f, Show1 g, Show a) => Show ((f :.: g) a) where showsPrec = showsPrec1
+#else
+-- kludge to get type with the same instances as g a
+newtype Apply g a = Apply (g a)
+
+getApply :: Apply g a -> g a
+getApply (Apply x) = x
+
+instance (Eq1 g, Eq a) => Eq (Apply g a) where
+    Apply x == Apply y = eq1 x y
+
+instance (Ord1 g, Ord a) => Ord (Apply g a) where
+    compare (Apply x) (Apply y) = compare1 x y
+
+instance (Read1 g, Read a) => Read (Apply g a) where
+    readsPrec d s = [(Apply a, t) | (a, t) <- readsPrec1 d s]
+
+instance (Show1 g, Show a) => Show (Apply g a) where
+    showsPrec d (Apply x) = showsPrec1 d x
+
+instance (Functor f, Eq1 f, Eq1 g, Eq a) => Eq ((f :.: g) a) where
+    Comp x == Comp y = eq1 (fmap Apply x) (fmap Apply y)
```

Comment:
this is copied from `transformers-0.4`, didn't give it a though. But there we have `class Eq1 f where eq1 :: Eq a => f a -> f a -> Bool`, and do use `Eq1 f` and `Eq1 g` instances, we have to wrap stuff in `Apply`.

This is one example why explicit dictionary passing style is better.